src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,255 @@
-// Saddleback search is a technique used to find an element in a sorted 2D matrix in O(m + n) time,
-// where m is the number of rows, and n is the number of columns. It works by starting from the
-// top-right corner of the matrix and moving left or down based on the comparison of the current
-// element with the target element.
 use std::cmp::Ordering;
 
-pub fn saddleback_search(matrix: &[Vec<i32>], element: i32) -> (usize, usize) {
-    // Initialize left and right indices
+/// Custom error type to represent errors related to matrix validation.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixError {
+    NonRectangularInput,
+    NotSorted,
+}
+
+/// Checks if the given matrix (vector of vectors) is sorted row-wise and column-wise.
+///
+/// # Arguments
+///
+/// * `matrix` - A vector of vectors representing the matrix to check.
+///
+/// # Returns
+///
+/// Returns `true` if the matrix is sorted both row-wise and column-wise. Otherwise, returns `false`.
+fn is_sorted(matrix: &[Vec<isize>]) -> bool {
+    let rows = matrix.len();
+    let cols = matrix[0].len();
```

Comment:
I think we need to check if the input is _rectangular_.

Reply:
We can move the logic that checks the matrix form and `is_sorted` into a single function, for example `validate_matrix`

Reply:
As above: I would introduce a new function checking if the input is _rectangular_ and use it in both public functions.